src/shims/native_lib/mod.rs

```diff
@@ -226,46 +217,56 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
             let evt_rg = evt.get_range();
             // LLVM at least permits vectorising accesses to adjacent allocations,
             // so we cannot assume 1 access = 1 allocation. :(
             let mut rg = evt_rg.addr..evt_rg.end();
             while let Some(curr) = rg.next() {
-                let alloc_id = this
-                    .alloc_id_from_addr(curr.to_u64(), rg.len().try_into().unwrap(), true)
-                    .expect("Foreign code did an out-of-bounds access!");
+                let Some(alloc_id) =
+                    this.alloc_id_from_addr(curr.to_u64(), rg.len().try_into().unwrap(), true)
+                else {
+                    throw_ub_format!("Foreign code did an out-of-bounds access!")
+                };
                 let alloc = this.get_alloc_raw(alloc_id)?;
                 let alloc_addr = alloc.get_bytes_unchecked_raw().addr();
 
                 // Skip forward however many bytes of the access are contained in the current allocation.
+                // The start of the overlap range will be the greater of the alloc base address and the
+                // lowest remaining address in the access' range; the end will be the lesser of the end
+                // of the allocation and the end of the access' range. Both are then shifted by `alloc_addr`
+                // so that the overlap range would have index 0 as the first byte of the allocation (which
+                // is how the methods on the allocation expect it to be).
                 let overlap = std::cmp::max(alloc_addr, curr).strict_sub(alloc_addr)
                     ..std::cmp::min(alloc_addr.strict_add(alloc.len()), rg.end)
                         .strict_sub(alloc_addr);
+                // Since `overlap` is capped in length by `rg.end`, this will inherently always be
+                // Ok(()) so we can discard it.
                 let _ = rg.advance_by(overlap.len());
-                //eprintln!("overlap: {overlap:?}");
-
-                // Get an overlap range as an offset from the allocation base addr.
-                //let overlap = unshifted_overlap.start..unshifted_overlap.start.strict_add(unshifted_overlap.len());
-
-                // Reads are infallible, writes might not be.
-                if evt.is_read() {
-                    let p_map = alloc.provenance();
-                    for idx in overlap {
-                        // If a provenance was read by the foreign code, expose it.
-                        if let Some(prov) = p_map.get(Size::from_bytes(idx), this) {
-                            this.expose_provenance(prov)?;
+
+                match evt {
+                    AccessEvent::Read(_) => {
+                        let p_map = alloc.provenance();
+                        for idx in overlap {
+                            // If a provenance was read by the foreign code, expose it.
+                            if let Some(prov) = p_map.get(Size::from_bytes(idx), this) {
```

Comment:
Please add a FIXME saying that we should really have a `get_range` on `ProvenanceMap`.

(Here's a next PR for you if you are interested. ;)